crates/transport-ipc/src/mock.rs

```diff
@@ -43,47 +147,347 @@ impl Default for MockIpcServer {
-    pub fn add_reply<S: Serialize>(&mut self, s: S) {
-        let reply = serde_json::to_vec(&s).unwrap();
-        self.add_raw_reply(reply);
+    /// Get the path to the Unix domain socket that this server will listen on.
+    /// This path should be passed to the provider being tested.
+    ///
+    /// # Returns
+    /// * `PathBuf` containing the socket path
+    pub fn path(&self) -> PathBuf {
+        self.inner.temp_file.path().to_owned()
     }
 
-    /// Add a json-rpc response to the server.
-    pub fn add_response<S: Serialize>(&mut self, response: Response<S>) {
-        self.add_reply(response);
+    /// Create a new handle to control this server.
+    /// The handle can be used to add responses and trigger shutdown.
+    ///
+    /// # Returns
+    /// * A new `MockIpcHandle` instance
+    pub fn handle(&self) -> MockIpcHandle {
+        MockIpcHandle { inner: self.inner.clone() }
     }
 
-    /// Get the path to the socket.
-    pub fn path(&self) -> PathBuf {
-        self.path.path().to_owned()
+    /// Handle a single client connection.
+    /// This function runs in a separate task for each connected client.
+    ///
+    /// # Arguments
+    /// * `stream` - The Unix domain socket stream for the client
+    /// * `inner` - Reference to shared server state
+    async fn handle_connection(
+        stream: tokio::net::UnixStream,
+        inner: Arc<Inner>,
+    ) -> std::io::Result<()> {
+        let (read, mut writer) = stream.into_split();
+        let mut reader = ReadJsonStream::new(read);
+
+        // Channel for sending responses back to the writer
+        let (tx, mut rx) = mpsc::channel::<Vec<u8>>(32);
+
+        // Collection of in-flight request handlers
+        let mut request_handlers = FuturesUnordered::new();
+
+        debug!("Starting connection handler loop");
+        loop {
+            tokio::select! {
+                // Handle incoming requests
+                maybe_request = reader.next() => {
+                    match maybe_request {
+                        Some(request) => {
+                            if let Ok(request) = serde_json::from_value::<Value>(request) {
+                                debug!(
+                                    id = ?request.get("id"),
+                                    method = ?request.get("method"),
+                                    "Received JSON-RPC request"
+                                );
+
+                                // Clone what we need for the task
+                                let inner = inner.clone();
+                                let tx = tx.clone();
+
+                                // Spawn a new task to handle this request
+                                request_handlers.push(tokio::spawn(async move {
+                                    // Get the next queued response or generate an error
+                                    let response = inner.replies.lock().pop_front().map_or_else(|| {
```

Comment:
this is also problematic because there's no guarantee that this task will pop the item that was at first position when the task was spawned